Refactor kinesis exporter to be aligned with Kafka Producer behaviour

MovieStoreGuy opened the pull request:
**Description:**
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.>
Updating the kinesis exporter to aligned with the behaviour of the Kafka exporter.

This removes the third party Kinesis Producer Library and  allows for the internal exporter pipeline to allow for things like retries and better reporting of data that is unable to be sent to Kinesis.

**Link to tracking Issue:** Currently a draft, will link the issue when it is ready.

**Testing:** Unit testing has been adding throughout where code has been updated.

**Documentation:** <Describe the documentation added.>

MovieStoreGuy commented on exporter/awskinesisexporter/factory.go:
```diff
@@ -45,59 +53,49 @@ func createDefaultConfig() config.Exporter {
+
+	return &Exporter{
+		producer: producer,
+		batcher:  batch.NewJaegerBatcher(),
```
Needs to accept batch size and record size.

MovieStoreGuy commented on exporter/awskinesisexporter/internal/batch/batch.go:
```diff
@@ -0,0 +1,131 @@
+	}
+	return b.addRaw(data, key)
+}
+
```
Potentially a need to add for `AddJson(value interface{}, key string)`, unsure for this moment. Will leave for a later PR if that is the case.

MovieStoreGuy commented on exporter/awskinesisexporter/internal/batch/batch.go:
```diff
@@ -0,0 +1,131 @@
+	return chunk
+}
+
+func min(a, b int) int {
```
Can not wait for generics to remove crap like this.

MovieStoreGuy commented on exporter/awskinesisexporter/internal/batch/encode_jaeger.go:
```diff
@@ -12,47 +12,49 @@
-type jaeger struct {
-	kinesis *awskinesis.Exporter
-}
+type jaeger struct{}
```
Need to update to store max allowed batch size and record size when creating a batch to forward to kinesis.

irisgve commented on exporter/awskinesisexporter/factory.go:
```diff
@@ -45,59 +53,49 @@ func createDefaultConfig() config.Exporter {
+	return createExporter(ctx, params, conf)
+}
+
+func createExporter(_ context.Context, params component.ExporterCreateSettings, conf config.Exporter) (*Exporter, error) {
```
To align this better with the Kafka exporter, should we consider using more of exporterhelper methods? i.e. for instantiating a new exporter we use `exporterhelper.NewXXXExporter`

MovieStoreGuy replied:
I am in for that, I need to look into detail what methods they are using and I can carry them over here.

irisgve replied:
iirc it always interfacing with retry and queueing configurations so defo worth checking out. Creating via the exporterhelper also enables the internal metrics for the exporter by default I believe

MovieStoreGuy commented:
My git habits got in the way so I have squashed my local and pushed it up here. Have also rebased with main so I am hoping the issue with linting is resolved.

irisgve commented on exporter/awskinesisexporter/config.go:
```diff
@@ -21,35 +21,16 @@ import (
 // AWSConfig contains AWS specific configuration such as awskinesis stream, region, etc.
```
Minor: let's add these configs in README i.e. https://github.com/atlassian-forks/opentelemetry-collector-contrib/tree/master/exporter/awskinesisexporter

irisgve commented on exporter/awskinesisexporter/README.md:
```diff
@@ -1,3 +1,43 @@
-To be added.
\ No newline at end of file
+
+The kinesis exporter currently exports Jaeger traces (dynamic encodings coming) to the configured kinesis stream.
```
Nice thanks for this

irisgve commented on exporter/awskinesisexporter/exporter.go:
```diff
@@ -16,32 +16,68 @@ package awskinesisexporter
+	_ component.LogsExporter    = (*Exporter)(nil)
+)
+
+func createExporter(c config.Exporter, log *zap.Logger) (*Exporter, error) {
```
This by default creates a trace exporter given that the batching uses `NewJaeger`. What's the plan to extend with the other types and does it make more sense to name it like so and create different exporters for each.
